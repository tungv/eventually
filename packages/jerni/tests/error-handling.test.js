const createJourney = require("../lib/createJourney2");
const mapEvents = require("../lib/mapEvents");
const makeServer = require("./makeServer");
const makeTestStore = require("./makeTestStore");
const makeTestLogger = require("./makeTestLogger");

test("onError without throwing", async () => {
  jest.setTimeout(300);
  expect.assertions(6);
  const { server } = await makeServer({
    ns: "test-error-handling",
    port: 19084,
  });

  const [logger, logs] = makeTestLogger();

  try {
    const store1 = makeTestStore(
      mapEvents({
        type_1: event => event.id,
        type_2: event => {
          throw new Error("inside handler");
        },
      }),
    );

    const store2 = makeTestStore(
      mapEvents({
        type_1: event => event.id,
        type_2: event => event.id,
      }),
    );

    const journey = createJourney({
      writeTo: "http://localhost:19084",
      stores: [store1, store2],
      logger,
      onError: (error, event, storeWithError) => {
        expect(error.message).toEqual("inside handler");
        expect(store1).toBe(storeWithError);
        expect(event).toEqual(
          expect.objectContaining({
            type: "type_2",
            payload: {},
          }),
        );

        // no throwing means skipping
      },
    });

    await journey.commit({ type: "type_1", payload: {} });
    await journey.commit({ type: "type_2", payload: {} });
    await journey.commit({ type: "type_1", payload: {} });

    const db = await store1.getDriver();
    const outputs = [];
    for await (const output of journey.begin()) {
      // only 2 possible events because we have filters by type
      outputs.push(output);
      if (db.length === 2) {
        expect(db).toEqual([1, 3]);
        break;
      }
    }

    expect(await store2.getDriver()).toEqual([1, 2, 3]);
    expect(logs.join("\n")).toMatchSnapshot();
  } finally {
    server.close();
    server.destroy();
    // console.log(logs.join("\n"));
  }
});

test("onError rethrown", async () => {
  jest.setTimeout(300);
  const { server } = await makeServer({
    ns: "test-error-handling",
    port: 19084,
  });

  const [logger, logs] = makeTestLogger(makeTestLogger.LEVEL_INFO);

  try {
    const store = makeTestStore(
      mapEvents({
        type_1: event => event.id,
        type_2: event => {
          throw new Error("inside handler");
        },
      }),
    );

    const journey = createJourney({
      writeTo: "http://localhost:19084",
      stores: [store],
      logger,
      onError: (error, event, storeWithError) => {
        throw new Error("cannot recover");
      },
    });

    await journey.commit({
      type: "type_1",
      payload: {},
      meta: { occurred_at: 16e11 },
    });
    await journey.commit({
      type: "type_2",
      payload: {},
      meta: { occurred_at: 16e11 },
    });
    await journey.commit({
      type: "type_1",
      payload: {},
      meta: { occurred_at: 16e11 },
    });

    const db = await store.getDriver();
    for await (const output of journey.begin()) {
      console.log(output);
    }

    expect(db).toEqual([1]);
    expect(logs.join("\n")).toMatchSnapshot();
  } finally {
    server.close();
    server.destroy();
  }
});
