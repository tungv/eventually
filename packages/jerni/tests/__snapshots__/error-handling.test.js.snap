// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onError rethrown 1`] = `
"[INF] start receiving data
[ERR] unrecoverable error happened while processing event #2
[ERR] {
  type: 'type_2',
  payload: {},
  meta: {
    occurred_at: 1600000000000,
    client: 'jerni',
    clientVersion: 'test'
  },
  id: 2
}
[ERR] inside handler
[INF] stop processing events"
`;

exports[`onError without throwing 1`] = `
"[DBG] journey.begin({{
  pulseCount: 200,
  pulseTime: 10,
  serverUrl: undefined,
  cleanStart: false
}})
[DBG] sending http request to: http://localhost:19084/subscribe
[DBG] headers {
  'last-event-id': '0',
  includes: 'type_1,type_2',
  'burst-count': 200,
  'burst-time': 10
}
[DBG] socket opened!
[INF] start receiving data
[DBG] handling events #1 - #3
[INF] skipped offending event #2
[DBG] done
[INF] stop processing events"
`;
